### PR TITLE
Introduce behat context modifiers to execute scenarios variations

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This setup that aims to provide a simple mechanism that would allow us to run our behat tests in multiple configurations: different languages, currencies, settings ...
| Type?         | improvement
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |
| How to test?  |  php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml --profile virtual_products

## Background

We have now a nice behat test suite that covers lot of scenarios, including cart prices computations. However this suite is run for one currency, one language. We would like to run it for a lot of different configurations: differenr currencies (EUR, USD ...), with virtual or not products ...

Of course we want to do this without having to copy-paste ou scenarios, in order to keep them maintenable.

## Solution

This PR introduces into key FeatureConcepts some flags that control settings, and have Modifiers that can switch these flags on and off. These feature flags are controlled by the behat [profile](http://behat.org/en/latest/user_guide/configuration.html#overriding-default-params) you use.

If you run tests using `default` profile, it will work with all modifier flags set to off. It runs all tests.

However, if you run the behat tests using the `virtual_products` profile, it will:
- turn the virtual_product modifier flag to yes, which will modify how the "I create a product" step behaves, now it will create virtual products instead of standard products
- only target the tests marked with the `@test-with-modifier-virtual-products` flag

## Want to test ?

Pull this branch, and run
```
php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml --profile virtual_products
```

## CI setup

We need to make Travis run
```
php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml
```
then
```
php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml --profile virtual_products
```

Basically, Travis will need to run the behat tests *as many times as there are profiles*

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13596)

### CODE REVIEW

File: tests/Integration/Behaviour/Features/Context/Modifier/ProductFeatureContextModifier.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * Scenarios marked with this tag will be run with virtual products instead of standard products.
+ */
+class ProductFeatureContextModifier implements BehatContext
+{
+    const PROFILE_VIRTUAL_PRODUCTS = 'virtual_products';
+
+    /** @BeforeScenario */
+    public function before(BeforeScenarioScope $scope)
+    {
+        $parameters = $scope->getEnvironment()->getSuite()->getSetting('parameters');
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Modifier/ProductFeatureContextModifier.php:44]: We retrieve the suite parameters and use it to find out which profile has been used to run the tests.

File: tests/Integration/Behaviour/behat.yml [hunk 2]
```diff
@@ -27,9 +29,18 @@ default:
...
                 features: Features/Scenario/Translation
             contexts:
                 - Tests\Integration\Behaviour\Features\Context\TranslationFeatureContext
                 - Tests\Integration\Behaviour\Features\Context\ModuleFeatureContext
+
+virtual_products:
+    suites:
+        cart:
+            parameters:
+                profile: 'virtual_products'
```
Comment (matks) [inline tests/Integration/Behaviour/behat.yml:47]: We override the parameter from suite `cart` in this profile

File: tests/Integration/Behaviour/behat.yml [hunk 3]
```diff
@@ -27,9 +29,18 @@ default:
...
+            filters:
+                tags: '@test-with-modifier-virtual-products'
```
Comment (matks) [inline tests/Integration/Behaviour/behat.yml:49]: For this profile, we require to run only tests flagged as `@test-with-modifier-virtual-products`

File: tests/Integration/Behaviour/Features/Context/Modifier/ProductFeatureContextModifier.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        if (false === array_key_exists('profile', $parameters)) {
```
Comment (PierreRambaud) [inline tests/Integration/Behaviour/Features/Context/Modifier/ProductFeatureContextModifier.php]:
```php
!empty($parameters['profile']) && $parameters['profile'] === self::PROFILE_VIRTUAL_PRODUCTS
```

File: tests/Integration/Behaviour/Features/Context/ProductFeatureContext.php [hunk 5]
```diff
@@ -71,6 +73,11 @@ public function getProductWithName($productName)
         return $this->products[$productName];
     }
 
+    public function activateCreateVirtualProductsModifier()
```
Comment (PierreRambaud) [inline tests/Integration/Behaviour/Features/Context/ProductFeatureContext.php]: Maybe add a method to toggle this flag instead?

**Comment by matks** [comment]
@tomlev I started using the modifier on relevant tests in `tests/Integration/Behaviour/Features/Scenario/Cart/Calculation/Currency/currency.feature`, I got some interesting results:
```
...
Expects 149.4, got 143 instead (RuntimeException)
...
Expects 41.9, got 37.4 instead (RuntimeException)
...
Expects 102.3, got 97.9 instead
```

Can you pull the branch, run `php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml --profile virtual_products` and tell me if these are relevant bug reports or I missed something ?

**Comment by tomlev** [comment]
as we discussed privately, this must be due to virtual cart (with only virtual products) has no shipping fees

**Comment by matks** [comment]
Caramba ! Encore raté !

File: tests/Integration/Behaviour/Features/Context/ProductFeatureContext.php [hunk 6]
```diff
@@ -188,6 +204,11 @@ protected function createProduct($productName, $price, $productQuantity)
         $product->price = $price;
         $product->name = $productName;
         $product->quantity = $productQuantity;
+
+        if ($this->flagModifierCreateVirtualProducts) {
+            $product->is_virtual = 1;
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/ProductFeatureContext.php:209]:
```suggestion
            $product->is_virtual = true;
```
